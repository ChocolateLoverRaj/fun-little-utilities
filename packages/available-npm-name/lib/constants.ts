export const PKG_NAME_RGX = /^((?:[\w._-]+)|(?:\@[\w._-]+\/[\w._-]+))$/
export const NPM_REGISTRY = 'https://registry.npmjs.org'
