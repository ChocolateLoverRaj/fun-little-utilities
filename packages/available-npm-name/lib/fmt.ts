import { Status } from './types'

interface StrMap {
  [_: string]: string
}

const prefix: StrMap = {
  [Status.NetworkError]: '❓',
  [Status.InvalidName]: '👎',
  [Status.Occupied]: '❌',
  [Status.Available]: '✅'
}

const suffix: StrMap = {
  [Status.NetworkError]: 'ran into a network error',
  [Status.InvalidName]: 'is an invalid name',
  [Status.Occupied]: 'is occupied',
  [Status.Available]: 'is available'
}

/** Convert a package name and a status into a message */
export const fmt = (
  packageName: string,
  status: Status
) => `${prefix[status]} ${packageName} ${suffix[status]}`

export default fmt
