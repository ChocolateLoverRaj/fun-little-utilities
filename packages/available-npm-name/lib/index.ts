export * from './constants'
export * from './fmt'
export * from './main'
export * from './parse-input'
export * from './types'
export * from './unit'
