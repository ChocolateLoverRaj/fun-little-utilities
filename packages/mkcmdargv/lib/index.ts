import { partition } from 'ramda'
import { splitIterable } from 'split-iterable'

type Value = string | number | boolean

type Options = {
  readonly [name in string]: Value
}

export interface Param {
  readonly args?: readonly string[]
  readonly options?: Options
  readonly flags?: readonly string[]
}

function parseSingleDash ([key, value]: [string, Value]) {
  const optName = '-' + key
  if (value === true) return [optName]
  if (value === false) return []
  return [optName, String(value)]
}

function parseDoubleDash ([key, value]: [string, Value]) {
  const optName = '--' + key
  if (value === true) return [optName]
  if (value === false) return []
  return [`${optName}=${value}`]
}

export function * iterateCommandArguments (param: Param) {
  const { args = [], options = {}, flags = [] } = param
  const [beforeDoubleDashes, ...afterDoubleDashes] = splitIterable(args, x => x === '--')
  const [beforeDashes, afterDashes] = partition(x => x[0] !== '-', beforeDoubleDashes)
  const optionsPairs = Object.entries(options)
  const [singleDashOptions, doubleDashOptions] = partition(([key]) => key.length === 1, optionsPairs)
  const [singleDashFlags, doubleDashFlags] = partition(x => x.length === 1, flags)

  yield * beforeDashes

  if (singleDashFlags.length) yield '-' + singleDashFlags.join('')

  for (const entry of singleDashOptions) {
    yield * parseSingleDash(entry)
  }

  yield * doubleDashFlags.map(x => '--' + x)

  for (const entry of doubleDashOptions) {
    yield * parseDoubleDash(entry)
  }

  if (!afterDashes.length && !afterDoubleDashes.length) return

  yield '--'
  yield * afterDashes

  for (const chunk of afterDoubleDashes) {
    yield '--'
    yield * chunk
  }
}

export function createCommandArguments (param: Param) {
  return Array.from(iterateCommandArguments(param))
}

export async function help () {
  const pathPromise = import('path')
  const fsxPromise = import('fs-extra')
  const path = await pathPromise
  const helpFilePath = path.resolve(__dirname, '../data/help.txt')
  const { readFile } = await fsxPromise
  return readFile(helpFilePath, { encoding: 'utf8' })
}
