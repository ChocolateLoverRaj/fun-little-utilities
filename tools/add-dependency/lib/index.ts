import path from 'path'
import { prompt } from 'inquirer'
import {
  Manifest,
  ManifestType,
  PackageManifest,
  TestManifest,
  ToolManifest,
  ManifestItem,
  ManifestList,
  NonRootManifestTypes,
  loadPackageList,
  loadTestList,
  loadToolList
} from '@tools/utils'

async function promptManifestType (): Promise<NonRootManifestTypes.Union> {
  const { type } = await prompt({
    name: 'type',
    message: 'Dependant container',
    type: 'list',
    choices: NonRootManifestTypes
      .map(value => ({ name: value.toLowerCase(), value }))
  })

  return type
}

interface GetVersionRequirementParam {
  readonly item: ManifestItem<PackageManifest>
  readonly manifest: PackageManifest
}

async function addProdDeps<Obj extends Manifest> (
  target: ManifestItem<Obj>,
  packages: ManifestList<PackageManifest>,
  getVersionRequirement: (param: GetVersionRequirementParam) => string
) {
  if (!packages.size) return

  const [
    manifest,
    dependencies
  ] = await Promise.all([
    target.readManifest(),
    Promise.all(packages.items().map(async item => ({
      item,
      manifest: await item.readManifest()
    })))
  ])

  if (!manifest.dependencies) {
    // @ts-ignore
    manifest.dependencies = {}
  }

  for (const param of dependencies) {
    const version = getVersionRequirement(param)
    // @ts-ignore
    manifest.dependencies[param.manifest.name] = version
  }

  await target.writeManifest(manifest)
}

export async function handlePackage () {
  const list = await loadPackageList()
  const target = await list.promptItem('Choose a package')
  const dependencies = await list.promptItemList('Choose dependencies')
  await addProdDeps(target, dependencies, param => '^' + param.manifest.version)
}

async function handleTestOrTool<Manifest extends TestManifest | ToolManifest> (
  loadTargetList: () => Promise<ManifestList<Manifest>>
) {
  const [
    targetList,
    packageList
  ] = await Promise.all([
    loadTargetList(),
    loadPackageList()
  ])

  const target = await targetList.promptItem('Choose a folder')
  const dependencies = await packageList.promptItemList('Choose packages')

  await addProdDeps(
    target,
    dependencies,
    param => path.relative(target.folder, param.item.folder)
  )
}

export async function handleTest () {
  await handleTestOrTool(loadTestList)
}

export async function handleTool () {
  await handleTestOrTool(loadToolList)
}

export async function main () {
  switch (await promptManifestType()) {
    case ManifestType.Package:
      return handlePackage()
    case ManifestType.Test:
      return handleTest()
    case ManifestType.Tool:
      return handleTool()
  }
}

export default main
