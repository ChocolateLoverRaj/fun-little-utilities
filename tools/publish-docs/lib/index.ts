import process from 'process'
import { publish, Options } from '@tools/gh-pages'
import places from '@tools/places'
import { optionsList } from './config'

const DEFAULT_OPTIONS: Options = { dotfiles: true }

export async function main () {
  process.chdir(places.project)
  console.info('Publishing documentation...')

  for (const options of optionsList) {
    console.info('gh-pages>', options)
    await publish('docs', { ...DEFAULT_OPTIONS, ...options })
  }
}

export default main
