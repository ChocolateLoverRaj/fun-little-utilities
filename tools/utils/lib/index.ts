export * from './esm'
export * from './logger'
