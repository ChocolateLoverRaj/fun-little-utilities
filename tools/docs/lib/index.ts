import path from 'path'
import process from 'process'
import { ensureDir, pathExists } from 'fs-extra'
import { Application } from 'typedoc'
import places from '@tools/places'
import { loadPackageList } from '@tools/utils'

export async function main () {
  const failures = []

  await ensureDir(places.docs)

  const list = await loadPackageList()
  for (const item of list.items()) {
    console.info('docs>', item.name)

    const outputDir = path.join(places.docs, item.name)

    const readme = path.join(item.folder, 'README.md')
    const readmeObject = await pathExists(readme) ? { readme } : null

    const app = new Application({
      target: 'esnext',
      module: 'commonjs',
      experimentalDecorators: true,
      logger: 'none',
      exclude: ['**/node_modules', '**/.git'],
      ...readmeObject
    })

    const project = app.convert(app.expandInputFiles([item.folder]))

    if (!project) {
      failures.push(item)
      continue
    }

    app.generateDocs(project, outputDir)
  }

  if (failures.length) {
    console.error()
    console.error('[ERROR]: Failed to generate docs for:')

    for (const item of failures) {
      console.error(' -', item.name)
    }

    throw process.exit(1)
  }
}

export default main
