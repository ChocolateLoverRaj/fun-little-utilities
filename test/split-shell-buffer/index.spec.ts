import EventEmitter from 'events'
import { normalText, styledText } from './.lib/data'

import {
  create,
  fromIterable,
  fromString,
  fromEventedStream,
  fromIterableStream,
  fromChildProcess,
  toString,
  write,
  writeln,
  StdOutError,
  StdErrError,
  SplitterObject,
  SequenceFunc,
  EventedStream,
  ChildProcess
} from 'split-shell-buffer'

const mkemit = (emitter: EventEmitter) => (timeout: number, event: string, ...args: any[]) => {
  setTimeout(() => emitter.emit(event, ...args), timeout)
}

class EventedStreamObject extends EventEmitter implements EventedStream {}

class ChildProcessObject extends EventEmitter implements ChildProcess {
  public readonly stdout = new EventedStreamObject()
  public readonly stderr = new EventedStreamObject()
}

describe('create()', () => {
  describe('constructs an object', () => {
    describe('without providing optional properties', () => {
      const splitter = create({
        data: Buffer.from(normalText)
      })

      it('which is an instance of SplitterObject', () => {
        expect(splitter).toBeInstanceOf(SplitterObject)
      })

      it('which has desired properties', () => {
        expect(splitter).toMatchObject({
          prefix: [],
          suffix: [],
          data: Buffer.from(normalText)
        })
      })
    })
  })

  describe('with prefix and suffix as instances of Buffer', () => {
    const data = Buffer.from(normalText)
    const prefix = Buffer.from('<prefix>')
    const suffix = Buffer.from('<suffix>')

    const splitter = create({ data, prefix, suffix })

    it('which is an instance of SplitterObject', () => {
      expect(splitter).toBeInstanceOf(SplitterObject)
    })

    it('which has desired properties', () => {
      expect(splitter).toMatchObject({ data, prefix, suffix })
    })
  })

  describe('with prefix and suffix as functions that return instances of Buffer', () => {
    type SeqParam = SequenceFunc.Param
    const data = Buffer.from(normalText)

    const createTracker = (collection: SeqParam[], data: Buffer) => (param: SeqParam) => {
      collection.push(param)
      return data
    }

    const prefixParamCollection = Array<SeqParam>()
    const prefix = createTracker(prefixParamCollection, Buffer.from('<prefix>'))

    const suffixParamCollection = Array<SeqParam>()
    const suffix = createTracker(suffixParamCollection, Buffer.from('<suffix>'))

    const splitter = create({ data, prefix, suffix })
    const stringPromise = toString(splitter)

    it('which is an instance of SplitterObject', () => {
      expect(splitter).toBeInstanceOf(SplitterObject)
    })

    it('which has desired properties', () => {
      expect(splitter).toMatchObject({ data, prefix, suffix })
    })

    it('calls prefix and suffix as functions with desired parameters', async () => {
      await stringPromise
      expect({ prefixParamCollection, suffixParamCollection }).toMatchSnapshot()
    })

    it('produces desired string', async () => {
      expect(await stringPromise).toBe([
        '<prefix>abc def ghi<suffix>',
        '<prefix>jkl mno pqrs<suffix>'
      ].join('\n'))
    })
  })
})

describe('fromIterable()', () => {
  it('constructs correct object', () => {
    expect(
      fromIterable(Buffer.from(normalText))
    ).toEqual(
      create({
        data: Buffer.from(normalText)
      })
    )
  })
})

describe('fromEventedStream()', () => {
  class Init {
    public readonly stream = new EventedStreamObject()
    public readonly splitter = fromEventedStream(this.stream)
    public readonly emit = mkemit(this.stream)
  }

  it('when provided stream emits all data at once', async () => {
    const { splitter, emit } = new Init()

    emit(100, 'data', Buffer.from(normalText))
    emit(200, 'close')

    expect(await toString(splitter)).toBe(normalText)
  })

  it('when provided stream emits data line by line', async () => {
    const { splitter, emit } = new Init()

    normalText.split('\n')
      .map(line => line + '\n')
      .map(line => Buffer.from(line))
      .forEach((line, index) => emit(index, 'data', line))

    emit(100, 'close')

    expect(await toString(splitter)).toBe(normalText + '\n')
  })

  it('when provided stream emits data character by character', async () => {
    const { splitter, emit } = new Init()

    Array
      .from(normalText)
      .map(char => Buffer.from(char))
      .forEach((char, index) => emit(index, 'data', char))

    emit(normalText.length, 'close')

    expect(await toString(splitter)).toBe(normalText)
  })

  it('when provided stream emits data as string', async () => {
    const { splitter, emit } = new Init()

    emit(100, 'data', normalText)
    emit(200, 'close')

    expect(await toString(splitter)).toBe(normalText)
  })

  it('when provided stream emits an error', () => {
    const error = new Error('Expected')
    const { splitter, emit } = new Init()

    emit(100, 'error', error)

    // tslint:disable-next-line:no-floating-promises
    expect(toString(splitter)).rejects.toBe(error)
  })
})

describe('fromIterableStream()', () => {
  it('when provided stream yields instances of Buffer', async () => {
    async function * iterate () {
      for (const chunk of normalText.split('\n')) {
        yield Buffer.from(chunk + '\n')
      }
    }

    const stream = {
      [Symbol.asyncIterator]: iterate
    }

    expect(
      await toString(fromIterableStream(stream))
    ).toBe(normalText + '\n')
  })

  it('when provided stream yields strings', async () => {
    async function * iterate () {
      for (const chunk of normalText.split('\n')) {
        yield chunk + '\n'
      }
    }

    const stream = {
      [Symbol.asyncIterator]: iterate
    }

    expect(
      await toString(fromIterableStream(stream))
    ).toBe(normalText + '\n')
  })
})

describe('fromChildProcess()', () => {
  class Init {
    public readonly process = new ChildProcessObject()
    public readonly splitter = fromChildProcess(this.process)
    public readonly emitProcess = mkemit(this.process)
    public readonly emitStdOut = mkemit(this.process.stdout)
    public readonly emitStdErr = mkemit(this.process.stderr)
  }

  it('when provided child process yields output through stdout', async () => {
    const { splitter, emitProcess, emitStdOut } = new Init()

    emitStdOut(100, 'data', Buffer.from(normalText))
    emitStdOut(200, 'close')
    emitProcess(200, 'close')

    expect(await toString(splitter)).toBe(normalText)
  })

  it('when provided child process yields output through stderr', async () => {
    const { splitter, emitProcess, emitStdErr } = new Init()

    emitStdErr(100, 'data', Buffer.from(normalText))
    emitProcess(200, 'close')

    expect(await toString(splitter)).toBe(normalText)
  })

  it('when provided child process yields output through both stdout and stderr', async () => {
    const { splitter, emitProcess, emitStdOut, emitStdErr } = new Init()

    emitStdOut(10, 'data', Buffer.from('out 0\n'))
    emitStdErr(20, 'data', Buffer.from('err 0\n'))
    emitStdOut(30, 'data', Buffer.from('out 1\n'))
    emitStdOut(40, 'data', Buffer.from('out 2\n'))
    emitStdErr(50, 'data', Buffer.from('err 1\n'))
    emitStdErr(60, 'data', Buffer.from('err 2\n'))
    emitProcess(100, 'close')

    expect(await toString(splitter)).toBe([
      'out 0',
      'err 0',
      'out 1',
      'out 2',
      'err 1',
      'err 2',
      ''
    ].join('\n'))
  })

  it('when stdout from provided child process emits an error', async () => {
    const { splitter, process, emitStdOut } = new Init()
    const expectedError = new Error('Expected')

    emitStdOut(100, 'error', expectedError)

    await toString(splitter).then(
      () => {
        throw new Error('Expecting a rejection but it resolves')
      },

      (error: StdOutError) => {
        expect(error).toBeInstanceOf(StdOutError)
        expect(error.error).toBe(expectedError)
        expect(error.stream).toBe(process.stdout)
        expect(error.name).toBe('StdOutError')
        expect(error.message).toBe(String(error.error))
      }
    )
  })

  it('when stderr from provided child process emits an error', async () => {
    const { splitter, process, emitStdErr } = new Init()
    const expectedError = new Error('Expected')

    emitStdErr(100, 'error', expectedError)

    await toString(splitter).then(
      () => {
        throw new Error('Expecting a rejection but it resolves')
      },

      (error: StdErrError) => {
        expect(error).toBeInstanceOf(StdErrError)
        expect(error.error).toBe(expectedError)
        expect(error.stream).toBe(process.stderr)
        expect(error.name).toBe('StdErrError')
        expect(error.message).toBe(String(error.error))
      }
    )
  })

  it('when provided child process emits an error', async () => {
    const { splitter, emitProcess } = new Init()
    const error = new Error('Expected')

    emitProcess(100, 'error', error)

    // tslint:disable-next-line:no-floating-promises
    expect(toString(splitter)).rejects.toBe(error)
  })
})

describe('fromString()', () => {
  it('constructs correct object', () => {
    expect(
      fromString(normalText)
    ).toEqual(
      create({
        data: Buffer.from(normalText)
      })
    )
  })
})

describe('toString()', () => {
  describe('with default options', () => {
    it('with normal text', async () => {
      expect(
        await toString(fromString(normalText))
      ).toEqual(
        await toString(
          fromString(normalText),
          { finalNewLine: false, encoding: 'utf8' }
        )
      )
    })

    it('with styled text', async () => {
      expect(
        await toString(fromString(styledText))
      ).toMatchSnapshot()
    })
  })

  describe('preserves normal text', () => {
    it('without options', async () => {
      expect(
        await toString(fromString(normalText))
      ).toEqual(normalText)
    })

    it('with finalNewLine = true', async () => {
      expect(
        await toString(
          fromString(normalText),
          { finalNewLine: true }
        )
      ).toEqual(normalText + '\n')
    })

    it('with encoding = "hex"', async () => {
      expect(
        await toString(
          fromString(normalText),
          { encoding: 'hex' }
        )
      ).toEqual(
        Buffer.from(normalText).toString('hex')
      )
    })
  })
})

describe('write()', () => {
  describe('with normal text', () => {
    const splitter = fromString(normalText)

    it('passes instances of Buffer to writable.write', async () => {
      await write({
        write (instance) {
          expect(instance).toBeInstanceOf(Buffer)
        }
      }, splitter)
    })

    it('passes buffers by line to writable.write', async () => {
      let count = 0

      await write({
        write () {
          count += 1
        }
      }, splitter)

      expect(count).toBe(normalText.split('\n').length)
    })

    it('passes buffers that resemble original text to writable.write', async () => {
      let text = ''

      await write({
        write (buffer) {
          text += buffer.toString()
        }
      }, splitter)

      expect(text).toBe(normalText)
    })
  })

  describe('with styled text', () => {
    const splitter = fromString(styledText)

    it('passes instances of Buffer to writable.write', async () => {
      await write({
        write (instance) {
          expect(instance).toBeInstanceOf(Buffer)
        }
      }, splitter)
    })

    it('passes buffers by line to writable.write', async () => {
      let count = 0

      await write({
        write () {
          count += 1
        }
      }, splitter)

      expect(count).toBe(styledText.split('\n').length)
    })

    it('passes buffers that resemble toString() to writable.write', async () => {
      let text = ''

      await write({
        write (buffer) {
          text += buffer.toString()
        }
      }, splitter)

      expect(text).toBe(await toString(splitter))
    })
  })
})

describe('writeln()', () => {
  describe('with normal text', () => {
    const splitter = fromString(normalText)

    it('passes instances of Buffer to writable.write', async () => {
      await writeln({
        write (instance) {
          expect(instance).toBeInstanceOf(Buffer)
        }
      }, splitter)
    })

    it('passes buffers by line to writable.write', async () => {
      let count = 0

      await writeln({
        write () {
          count += 1
        }
      }, splitter)

      expect(count).toBe(normalText.split('\n').length)
    })

    it('passes buffers that resemble original text with trailing eol to writable.write', async () => {
      let text = ''

      await writeln({
        write (buffer) {
          text += buffer.toString()
        }
      }, splitter)

      expect(text).toBe(normalText + '\n')
    })
  })

  describe('with styled text', () => {
    const splitter = fromString(styledText)

    it('passes instances of Buffer to writable.write', async () => {
      await writeln({
        write (instance) {
          expect(instance).toBeInstanceOf(Buffer)
        }
      }, splitter)
    })

    it('passes buffers by line to writable.write', async () => {
      let count = 0

      await writeln({
        write () {
          count += 1
        }
      }, splitter)

      expect(count).toBe(styledText.split('\n').length)
    })

    it('passes buffers that resemble toString() with trailing eol to writable.write', async () => {
      let text = ''

      await writeln({
        write (buffer) {
          text += buffer.toString()
        }
      }, splitter)

      expect(text).toBe(await toString(splitter) + '\n')
    })
  })
})

describe('SplitterObject::withPrefix()', () => {
  it('creates a new instance', () => {
    const a = fromIterable([])
    const b = a.withPrefix([])
    expect(b).not.toBe(a)
  })

  it('creates an instance with modified prefix', () => {
    const prefix = Buffer.from('prefix')

    expect(
      fromIterable([])
        .withPrefix(prefix)
    ).toEqual(
      create({ data: [], prefix })
    )
  })
})

describe('SplitterObject::withSuffix()', () => {
  it('creates a new instance', () => {
    const a = fromIterable([])
    const b = a.withSuffix([])
    expect(b).not.toBe(a)
  })

  it('creates an instance with modified suffix', () => {
    const suffix = Buffer.from('prefix')

    expect(
      fromIterable([])
        .withSuffix(suffix)
    ).toEqual(
      create({ data: [], suffix })
    )
  })
})

describe('SplitterObject::withIndent()', () => {
  it('calls withPrefix()', () => {
    expect(
      fromIterable([])
        .withIndent(4)
    ).toEqual(
      fromIterable([])
        .withPrefix(Buffer.from(' '.repeat(4)))
    )
  })
})
