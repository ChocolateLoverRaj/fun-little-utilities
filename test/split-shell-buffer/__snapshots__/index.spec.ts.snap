// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toString() with default options with styled text 1`] = `
"abc [1mdef [2m
[0m[1m[2mghi [0;0mjkl [91;22;3m
[0m[91;22;3mmno pqrs [m
tuv wxyz
01 [2m23 [1m45
[0m[2m[1m67 [m89
 -- [4munderline[0] -- [24m
[0m[4m[24m[4m -- underline[1][24m -- 
[0m[4m[24m -- [7minverse[0] -- [27m
[0m[7m[27m[7m -- inverse[1][27m -- 
[0m[7m[27m -- [4m[7munderline.inverse[0] -- [27m[24m
[0m[4m[7m[27m[24m[4m[7m -- underline.inverse[1][27m[24m -- 
[0m[4m[7m[27m[24mfoo
bar"
`;
